File: 2334-number-of-flowers-in-full-bloom/number-of-flowers-in-full-bloom.py
class Solution:
    def fullBloomFlowers(self, flowers: List[List[int]], people: List[int]) -> List[int]:
        """Calculates the number of flowers that are in full bloom at a given time.

        Args:
            flowers: A list of lists of integers, where each inner list contains the start and end times of a flower's bloom period.
            people: A list of integers, where each integer represents the time at which a person visits the garden.

        Returns:
            A list of integers, where each integer represents the number of flowers that are in full bloom at the time of the corresponding person's visit.
        """

        # Sort the start and end times of the flower bloom periods.
        start = sorted([f[0] for f in flowers])
        end = sorted([f[1] for f in flowers])

        # Calculate the number of flowers that are in full bloom at the time of each person's visit.
        ans = []
        for p in people:
            blommed_till_now = bisect.bisect_right(start, p)
            bloomed_till_off = bisect.bisect_left(end, p)
            ans.append(blommed_till_now - bloomed_till_off)

        return ans

File: 2334-number-of-flowers-in-full-bloom/number-of-flowers-in-full-bloom.cpp
class Solution {
public:
    vector<int> fullBloomFlowers(vector<vector<int>>& flowers, vector<int>& people) {
        vector<int> ans;
        vector<int> start;
        vector<int> end; 

        for(int i=0; i<flowers.size(); i++) {
            start.push_back(flowers[i][0]);
            end.push_back(flowers[i][1]);
        }

        sort(start.begin(), start.end());
        sort(end.begin(), end.end());

        for (int i = 0; i < people.size(); i++) {
            int blommed_till_now = upper_bound(start.begin(), start.end(), people[i]) - start.begin();
            int bloomed_till_off = lower_bound(end.begin(), end.end(), people[i]) - end.begin();
            ans.push_back(blommed_till_now - bloomed_till_off);
        }

        return ans; 
    }
};